### PR TITLE
fastify-session integration

### LINKED ISSUE #319: Support @fastify/session out of the box
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

As https://github.com/fastify/session became part of the org, we should list it here as one of the default session manager. I think it would _just_ work, but it still needs tests and docs.

### Motivation

_No response_

### Example

_No response_

### LINKED ISSUE #376: `state` for passport causes `fastify-passport` to fail
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.22.1

### Plugin version

0.4.3

### Node.js version

16.x

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

macOS Big Sure 11.3.1

### Description

When I am trying to pass some encoded state to the strategy the callback URLs shows the user as unauthenticated, it's also unclear how I could pass the encoded state from the redirect endpoint dynamically.

### Steps to Reproduce

```typescript
fastifyPassport.use(
    "github",
    new Github.Strategy(
      {
        clientID: "xxx",
        clientSecret: "xxx",
        callbackURL: "http://lvh.me:3000/callback",
        scope: ["user:email", "user,user:email"],
        userProfileURL: "https://api.github.com/user",
        state: Buffer.from(JSON.stringify({ returnTo: "/mypage" })).toString(
          "base64"
        ),
        passReqToCallback: true,
      },
      function (req, accessToken, refreshToken, profile, done) {
        done(null, { ...profile, accessToken, query });
      }
    )
  );

server.get(
    "/github",
    {
      preValidation: fastifyPassport.authenticate("github", {
        successRedirect: "/callback",
        authInfo: false,
        assignProperty: "user",
      }),
    },
    () => {}
  );
  
  server.get(
  "/callback",
  {
    preValidation: fastifyPassport.authenticate("github", {
      authInfo: false,
      session: false,
    }),
  },
  async (request, reply, err, user = {}, info = {}, status = {}) => {
    reply.send({ user: request.user || {} });
  }
);

  ```

### Expected Behavior

Step through the prevalidation hook and return the encoded state to the endpoint.

### PR DESCRIPTION
**Pull request by sameer-coder**
This PR adds support for `fastify-session`.

PR includes:
- A new SessionManager class for fastify-session
- Tests
- Documentation update

Closes #319 
Closes #376

### CODE REVIEW

File: README.md
Comment (mcollina) [inline README.md:53]: Adding a link would be great. What's the difference between @fastify/session and fastify-secure-session? Can you document it?

File: src/AuthenticationRoute.ts
Comment (mcollina) [inline src/AuthenticationRoute.ts:16]: Do a typeof check, otherwise a get property in the session would crash it.

File: src/AuthenticationRoute.ts
Comment (mcollina) [inline src/AuthenticationRoute.ts:19]: Do a typeof check, otherwise a set property in the session would crash it.

File: src/Authenticator.ts
Comment (mcollina) [inline src/Authenticator.ts:32]: Can we use an interface instead?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: README.md
Comment (sameer-coder) [reply README.md]: I have added the link and some differences. Please review to see if it makes sense.

File: src/AuthenticationRoute.ts
Comment (airhorns) [inline src/AuthenticationRoute.ts:15]: I think it would make sense to add new functions to the session managers that know how to do this stuff instead of adding other wrapper functions out here. That way, to add a third supported library, folks just have to implement the session manager interface instead of editing these wrappers also. We could also then accept a session manager as a plugin option and there'd be no hardcoded bits that might need to be changed.

File: src/Authenticator.ts
Comment (airhorns) [inline src/Authenticator.ts:27]: I think this should have the name `SessionManager` and the class should have the name `FastifySessionManager` or something like that?

File: test/fastify-session/passport.test.ts
Comment (airhorns) [inline test/fastify-session/passport.test.ts:459]: Can we avoid duplicating all these tests, and instead wrap them up in a function that gets invoked twice in the appropriate different contexts? Its a bunch more maintenance burden to have to edit 2x as many tests, and other contributors might not know they need to add tests to both places.

**Comment by simoneb** [comment]
about the build failing, I don't know what the problem is but I would try removing the lockfile. having a lockfile which is by its nature not portable across node versions doesn't seem like a good idea. also in general we don't use lock files in repos containing packages

**Comment by airhorns** [comment]
@simoneb we had a big discussion about the lockfile here https://github.com/fastify/fastify-passport/pull/24, I think there are good reasons for keeping it, if you look at the list of failing depfu PRs on this repo right now they're all things that might break for users or break CI without the authors understanding why.

**Comment by simoneb** [comment]
> @simoneb we had a big discussion about the lockfile here #24, I think there are good reasons for keeping it, if you look at the list of failing depfu PRs on this repo right now they're all things that might break for users or break CI without the authors understanding why.

well we have this PR and we don't understand why it's failing either :)

File: src/AuthenticationRoute.ts
Comment (mcollina) [reply src/AuthenticationRoute.ts:15]: I agree

**Comment by airhorns** [comment]
Good thing we can rule out developers using different versions of the packages dependencies as a cause then! I don't think we should randomly pull levers to see if it fixes this PR, we should look at the fail! Here's the failure message:

```
  ● Test suite failed to run

    test/independent-strategy-instances.test.ts:26:40 - error TS2339: Property 'get' does not exist on type 'Session'.

    26     async (request) => request.session.get('messages')
                                              ~~~
    test/independent-strategy-instances.test.ts:69:53 - error TS2339: Property 'get' does not exist on type 'Session'.
 
    69     async (request) => `messages: ${request.session.get('messages')}`
                                                           ~~~
    test/independent-strategy-instances.test.ts:112:53 - error TS2339: Property 'get' does not exist on type 'Session'.

    112     async (request) => `messages: ${request.session.get('messages')}`
```

My guess is that one of the changes in this PR changes the type of the `request.session` property to be whichever is imported last, which is probably `@fastify/session`, overriding the interface that used to be in place before. Those tests could be updated to cast to the right session type, or to use the new helpers being added to the session manager to access the session.

File: src/AuthenticationRoute.ts
Comment (sameer-coder) [reply src/AuthenticationRoute.ts:15]: Makes sense. I have opened a [PR ](https://github.com/fastify/session/pull/50) on `@fastify/session` to add the new functions.
Correct me if I am wrong, once we have those function we wouldn't need any changes here in `fastify-passport`.  I will only need to update the documentation and add a reusable suite of tests which can be invoked with different session plugins.

Comment (climba03003) [reply src/AuthenticationRoute.ts:15]: I think he means the interface inside this plugin.

```js
class SessionManager {
  getSession(session: any, key: string) {
    return session[key];
  }
  
  setSession(session: any, key: string, value: any) {
    session[key] = value;
  }
}

class SecureSessionManager {
  getSession(session: any, key: string) {
    return session.get(key);
  }
  
  setSession(session: any, key: string, value: any) {
    session.set(key, value)
  }
}
```

Comment (sameer-coder) [reply src/AuthenticationRoute.ts:15]: > I think he means the interface inside this plugin.
> 
> ```js
> class SessionManager {
>   getSession(session: any, key: string) {
>     return session[key];
>   }
>   
>   setSession(session: any, key: string, value: any) {
>     session[key] = value;
>   }
> }
> 
> class SecureSessionManager {
>   getSession(session: any, key: string) {
>     return session.get(key);
>   }
>   
>   setSession(session: any, key: string, value: any) {
>     session.set(key, value)
>   }
> }
> ```

@climba03003  yeah probably he means inside the plugin but then that would add get/set methods to the `request.passport.sessionManager` object instead of `request.session`.  Adding those methods on @fastify/session would make it available on `request.session`.

File: test/decorators.test.ts
Comment (mcollina) [inline test/decorators.test.ts:55]: It's better to skip the test instead

File: test/decorators.test.ts
Comment (mcollina) [inline test/decorators.test.ts:36]: It's better to skip the test instead

**Review by mcollina — COMMENTED** [review]
amazing work!

File: test/run.js
Comment (mcollina) [inline test/run.js:9]:
```suggestion
  console.log('Running tests with session', sessionPlugin, '\n')
  const result = await runCLI(jestConfig, ['./'])
```

File: test/run.js
Comment (mcollina) [inline test/run.js:11]:
```suggestion
    console.log(`Tests completed\n`)
```

File: test/run.js
Comment (mcollina) [inline test/run.js:13]: I think we should stop the execution of the tests maybe?

Comment (sameer-coder) [reply test/run.js]: Not sure. Maybe we can let it to run so you can see failed tests from both runs?

File: test/decorators.test.ts
Comment (sameer-coder) [reply test/decorators.test.ts:55]: unlike tap jest doesn't allow you to skip the test once it has started running 😕 . To get it to skip I might have to add some more conditional logic before the test runs.

Comment (sameer-coder) [reply test/decorators.test.ts:55]: The community has been [asking ](https://github.com/facebook/jest/issues/8604) for a `skip` feature for a long time but jest team does not think skip should ever be necessary :)

**Review by simoneb — APPROVED** [review]

File: test/decorators.test.ts
Comment (airhorns) [reply test/decorators.test.ts:55]: You can't skip the test once it has started, but you can skip it before it starts with `test.skip("the test name")`. You could do `const sessionOnlyTest = process.env.SESSION_PLUGIN == "@fastify/session" ? test : test.skip` for a nice reusable primitive.

File: test/run.js
Comment (airhorns) [inline test/run.js:25]: Do you think we could avoid these non-standard scripts to make it easy for new contributors to know how to run the tests? This kinda stuff is generally unexpected and weird such that most folks wouldn't guess to look for it, and it also couples to jests internals a bit. Jest test and tap tests are just global function calls -- so couldn't we just use a function to duplicate the test executions?

```typescript
const suite = (sessionPluginName: string) => {
  describe(`${sessionPluginName} tests`, () => {
    test("one thing", () => { ... })
    test("another thing", () => { ... })
  })
}

suite("@fastify/session");
suite("fastify-secure-session");
```

That way, all the nice jest features like test timing, the log output, regex filtering, etc all just work because the test cases are all normal test cases, not part of some outer layer of double execution with environment variables.

File: test/session-serialization.test.ts
Comment (airhorns) [inline test/session-serialization.test.ts:139]: Jest has an `await expect(somePromise).rejects.toContain` helper if you want to make this a bit less verbose, the jest eslint rules recommend this form

File: package-lock.json
Comment (airhorns) [inline package-lock.json]: Why the downgrade here? I think you might need to update your version of npm

**Review by airhorns — CHANGES_REQUESTED** [review]
Can we not remove package-lock.json

File: test/run.js
Comment (sameer-coder) [reply test/run.js]: I implemented the way you described it. Seems much cleaner now. Thanks!

**Review by mcollina — APPROVED** [review]
lgtm

File: package-lock.json
Comment (airhorns) [inline package-lock.json]: Did you intend to upgrade all these dependencies? Any particular reason?

Comment (sameer-coder) [reply package-lock.json]: Nope, not intentional. I just did `npm i` and added the `@fastify/session` lib

Comment (zekth) [reply package-lock.json]: Not sure if we should keep the package-lock.

Comment (darkgl0w) [reply package-lock.json]: +1, It should be removed from the repo and added to the `.gitignore` file

Comment (sameer-coder) [reply package-lock.json]: I have removed the  package-lock

Comment (airhorns) [reply package-lock.json]: @zekth @darkgl0w have you read the big discussion we had about keeping it in #24 ? I still think it is a good idea to have around to know when upstream dependencies have broken the build. To reiterate, I am fully aware that the package lock has no impact on folks installing this package from NPM, but I do think that enforcing a consistent set of dependency versions for the different developers of this project and CI is worth something, just the same way as it would be for an application using this package. Why remove it?

Comment (zekth) [reply package-lock.json]: > the package lock has no impact on folks installing this package from NPM

So why not having the same behavior as the users of the project? Usually the goto is to freeze your versions for your end usage like applications/servers and such. Regarding lib this has been debated in all the possible ways you can imagine. Just this is the only repo i think which has a `package-lock.json` in the fastify-plugin ecosystem.

Comment (airhorns) [reply package-lock.json]: > So why not having the same behavior as the users of the project?

I think we should try to reveal what will happen to a new user who installs the project I agree -- it should work great out of the box with a fresh set of dependencies locked by someone else. But, I don't think a lockfile prevents that -- I think it actually helps, because of depfu. Without a package lock or depfu, the main branch of CI just starts failing all of a sudden when an upstream dependency breaks something. Or, worse, if there is no activity on the main branch, it will fail the next time but no has has been alerted to that yet. Instead, I'd rather have depfu be doing the bumps to track the most recent versions of everything, and get each bump to have CI run with github action's free minutes so we actually know if the dependency bump works. Then, if something breaks, we get a red pr where we can collaborate to fix it instead of it happening to some unsuspecting contributor trying to do something else completely. It's happened a couple times already with TypeScript on this repo where a new version of TS was released that breaks our build -- wouldn't you rather know about that as soon as it happens instead of having some user of the package waste their time discovering that and debugging it?

Comment (zekth) [reply package-lock.json]: Well i don't have strong opinion it's more for ecosystem consistency.

Sidenote: #413 issue is caused by Jest which we should avoid. Node-tap has a pretty good TS integration.

Comment (airhorns) [reply package-lock.json]: @sameer-coder just to unblock you, can you not remove the package-lock.json file from this PR, and the package maintainers can discuss that in a separate issue? That way you are unblocked to get this awesome change in!

Comment (darkgl0w) [reply package-lock.json]: @airhorns > as you suggested we should keep going on this PR and open a separate issue regarding the `package-lock.json` file removal and the use of jest in this repo.

Comment (sameer-coder) [reply package-lock.json:48]: @airhorns @darkgl0w I have added the package-lock back and rebased this PR. If everything looks good then let's merge it!

**Review by darkgl0w — CHANGES_REQUESTED** [review]
@sameer-coder > CI currently fails due to npm lock file version conflicts

**Review by darkgl0w — APPROVED** [review]
LGTM

**Review by zekth — APPROVED** [review]
✌️

**Review by airhorns — APPROVED** [review]

**Comment by sameer-coder** [comment]
I think this PR should be good to merge.

**Pull request merged**
